powersimdata/network/zenodo.py

```diff
@@ -0,0 +1,134 @@
+import hashlib
+import json
+import os
+import shutil
+from contextlib import contextmanager
+from zipfile import ZipFile
+
+import requests
+import wget
+
+url = "https://zenodo.org/api/records/"
+
+
+class Zenodo:
+    """Get data from a Zenodo archive.
+
+    :param str record_id: zenodo record id
+    """
+
+    def __init__(self, record_id):
+        """Constructor"""
+        self.record_id = record_id
+        self.content = self._get_record_content()
+
+    def _get_record_content(self):
+        """Make HTTP request to zenodo API and retrieve content.
+
+        :return: (*dict*) -- content of the response in unicode.
+        :raises Exception: if connection times out.
+        :raises ValueError: if record is invalid.
+        """
+        try:
+            r = requests.get(url + self.record_id, timeout=10)
+        except requests.exceptions.ConnectTimeout:
+            raise Exception("Connection to zenodo.org timed out")
+
+        if not r.ok:
+            raise ValueError("Record could not be accessed")
+        else:
+            content = json.loads(r.text)
+            metadata = content["metadata"]
+            print(f"Title: {metadata['title']}")
+            print(f"Publication date: {metadata['publication_date']}")
+            print(f"Version: {metadata['version']}")
+            print(f"DOI: {metadata['doi']}")
+
+            return content
+
+    def _get_remote_checksum(self, f):
+        """Get checksum of local copy of a file
+
+        :param dict f: dictionary containing information on the remote copy of a file.
+        :return: (*str*) -- checksum
+        """
+        return f["checksum"].split(":")[1]
+
+    def _get_local_checksum(self, f):
+        """Get remote copy of a file.
+
+        :param dict f: dictionary containing information on the remote copy of a file.
+        :return: (*str*) -- checksum if file exists
+        """
+        filename = os.path.join(self.dir, f["key"])
+        if not os.path.exists(filename):
+            return "invalid"
+        else:
+            h = hashlib.new(f["checksum"].split(":")[0])
+            with open(filename, "rb") as file:
+                bytes = file.read()
+            h.update(bytes)
+            return h.hexdigest()
+
+    @contextmanager
+    def _change_dir(self):
+        work_dir = os.getcwd()
+        os.chdir(os.path.expanduser(self.dir))
+        try:
+            yield
+        finally:
+            os.chdir(work_dir)
+
+    def _download_data(self, f):
+        """Fetch data.
+
+        :param dict f: information on the file to download.
+        """
+        with self._change_dir():
```

Comment:
Given `delete` ->`download` -> `unzip` are called sequentially, maybe we could put the decorator `with self._change_dir()` outside to call it once in `load_data`.

Reply:
Done